### ISSUE TITLE
Mesh secured Nomad jobs no longer possible across Datacenters?

### ISSUE THREAD

**Issue by neuroserve** [hashicorp/nomad#23505]
### Nomad version

Nomad v1.8.1
BuildDate 2024-06-19T06:43:57Z
Revision 5022543e4b7b8dcec9df123f86630ae3fdcffbe6

### Operating system and Environment details

All Linux (Debian 11) environment with federated Nomad regions on top of federated Consul datacenters in two separate cloud providers (not connected via vpn only via connect mesh gateways).

```
 +----------------+               +----------------+
 |                |               |                |
 | Nomad Region 1 |---federated---| Nomad Region 2 |
 |                |               |                |
 +----------------+               +----------------+
         |                                 |
 +------------------+           +------------------+
 |                  |           |                  |
 | Consul Cluster 1 |-federated-| Consul Cluster 2 |
 |                  |           |                  |
 +------------------+           +------------------+

 Cloud Provider 1                   Cloud Provider 2

```

### Issue

When trying to run a connect-enabled job, the job will not start. Instead I get the error
`failed to derive SI token: Unexpected response code: 500 (rpc error making call: Local tokens are disabled)`

### Reproduction steps
1. Set up at least two Consul federated datacenters with mesh gateways (that involves activating ACLs, token replication and enabling connect).
2. Set up at least two federated Nomad regions/clusters using the Consul datacenters created before (s. diagram)
3. Try to run a connect-enabled Job 

#### Expected Result

Job should start while Consul creates an SI token in the name of Nomad.

#### Actual Result

Job does not start. Instead you get the error `failed to derive SI token: Unexpected response code: 500 (rpc error making call: Local tokens are disabled)`. 
This article (https://support.hashicorp.com/hc/en-us/articles/17767098217235-How-Nomad-Manages-ACL-Tokens-Polices-for-Consul-Service-Mesh) tells us, that ACL tokens automatically generated by Nomad on behalf of Connect-enabled services are now created in [Local](https://developer.hashicorp.com/consul/docs/security/acl/acl-tokens#token-attributes) rather than Global scope, and are no longer replicated globally.

### Job file (if appropriate)
```
job "egw" {

  datacenters = ["dc1"]
  region = "de-west"
  node_pool = "static-clients"

  group "egw" {
    count = 2
    network {
      mode = "bridge"
    }

    service {
      name = "egw"
      connect {
        gateway {
          proxy {}
          terminating {
            service {
              name = "postgres"
            }
            service {
              name = "redis-svc"
            }
          }
        }
      }
    }
  }
}
```
This job is from https://medium.com/navin-nair/practical-hashicorp-nomad-and-consul-a-little-more-than-hello-world-part-1-991d2a54fd64

**Comment by pkazmierczak** [comment]
hi @neuroserve, thanks for reporting this. Just to clarify: do your Nomad agents use [Workload Identity](https://developer.hashicorp.com/nomad/docs/concepts/workload-identity), or just a Consul token to register with Consul?

**Comment by neuroserve** [comment]
The setup is still token based. I was already wondering, whether it should be switched to workload identity authentication. But when I read, I had to setup a loadbalancer in front of the Nomad nodes for production, I postponed that.

**Comment by neuroserve** [comment]
I can create Local tokens in the Consul GUI, but they can only be seen in the primary Consul cluster - not in the secondaries.

**Comment by neuroserve** [comment]
Consul Version is:
```
root@consul-0:~# consul version
Consul v1.17.3
Revision 009041f8
Build Date 2024-02-13T18:30:29Z

```

Consul info is:

```
root@consul-0:~# consul info
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 0
build:
        prerelease = 
        revision = 009041f8
        version = 1.17.3
        version_metadata = 
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 4
        leader = false
        leader_addr = 192.168.0.211:8300
        server = true
raft:
        applied_index = 677072
        commit_index = 677072
        fsm_pending = 0
        last_contact = 18.030745ms
        last_log_index = 677072
        last_log_term = 119
        last_snapshot_index = 671947
        last_snapshot_term = 119
        latest_configuration = [{Suffrage:Voter ID:03e09c98-016e-8899-9e82-3e3d69078069 Address:192.168.0.211:8300} {Suffrage:Voter ID:9f6f1ef9-67fc-ce4a-0653-3e2ec5119e13 Address:192.168.0.88:8300} {Suffrage:Voter ID:fd5d2820-7e0f-d9bf-d096-41e1e6662235 Address:192.168.0.66:8300}]
        latest_configuration_index = 0
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 119
runtime:
        arch = amd64
        cpu_count = 2
        goroutines = 283
        max_procs = 2
        os = linux
        version = go1.21.7
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 48
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 164
        members = 11
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 920
        members = 12
        query_queue = 0
        query_time = 1

```

**Comment by tgross** [comment]
Hi @neuroserve, I believe you're going to want `enable_token_replication = true` in your Consul ACL configuration. See https://github.com/hashicorp/nomad/issues/23728 for an example of someone who ran into the same issue. We likely need to update our documentation to reflect this for Workload Identity users.

**Comment by neuroserve** [comment]
As documented in https://github.com/hashicorp/nomad/issues/23728 I have set up a new environment (s above) with Consul 1.19.1 and Nomad 1.8.3.

First I tried to use `enable_token_replication = true` in both primary and secondary Consul clusters but that didn't work (neither from the start nor after the federation). Only setting it in the primary (and setting the replication token in the secondary) gives me a stable replication.
Tokens with global scope are replicated and can be created on the primary as well as the secondary.

Running the countdash-job

```
job "countdash" {
  datacenters = ["prod1"]
  node_pool = "static-clients"

  group "api" {
    network {
      mode = "bridge"
    }

    service {
      name = "count-api"
      port = "9001"

      connect {
        sidecar_service {}
      }
    }

    task "web" {
      driver = "docker"

      config {
        image = "hashicorpdev/counter-api:v3"
      }
    }
  }

  group "dashboard" {
    network {
      mode = "bridge"

      port "http" {
        static = 9002
        to     = 9002
      }
    }

    service {
      name = "count-dashboard"
      port = "http"

      connect {
        sidecar_service {
          proxy {
            upstreams {
              destination_name = "count-api"
              local_bind_port  = 8080
            }
          }
        }
      }
    }

    task "dashboard" {
      driver = "docker"

      env {
        COUNTING_SERVICE_URL = "http://${NOMAD_UPSTREAM_ADDR_count_api}"
      }

      config {
        image = "hashicorpdev/counter-dashboard:v3"
      }
    }
  }
}

```
in the primary datacenter (here 'prod1') works without problems. As soon as I change the datacenter to one of the secondaries ('prod4' in this case), I can no longer start it (the envoy proxy cannot be started). Error message is:

`Task hook failed: envoy_bootstrap: error creating bootstrap configuration for Connect proxy sidecar: exit status 1; see: <https://developer.hashicorp.com/nomad/s/envoy-bootstrap-error>`

Envoy logs:

`failed fetch proxy config from local agent: Unexpected response code: 403 (Permission denied: token with AccessorID 'primary-dc-down' lacks permission 'service:read' on "count-api-sidecar-proxy")`

Thus the previous error 

`failed to derive SI token: Unexpected response code: 500 (rpc error making call: Local tokens are disabled)`

is no longer there. 

I can see SI tokens, that Nomad creates, in the 'consul acl token list'. But they have a local scope and are therefore not replicated by the Consul ACL replication.

Am I missing something or is it no longer possible to run service mesh protected services within nomad and acreoss seperate cloud providers?
Can someone explain how https://github.com/hashicorp/consul/issues/7381 was resolved?
Do I have to change my config somewhere (what is the 'primary-dc-down' token and how do I add permissions to it)?

Any help would be appreciated.

**Comment by neuroserve** [comment]
When I split the above job in two (one for the api and one for the dashboard) and run one in prod1 and one in prod4, only the one in prod1 (primary consul) gets healthy. The job in prod4 (secondary consul) fails with the above errors.

**Comment by tgross** [comment]
Wanted to follow up here after an internal discussion on https://github.com/hashicorp/consul/issues/21863. Apparently the right way to fix this is to set token locality to global in the Consul auth method. But we don't necessarily recommend this, because it means that creating those Consul tokens is contingent on the primary datacenter being live during allocation startup. You'll need to decide for yourself whether that makes sense in your deployment environment.

**Comment by tgross** [comment]
Added a docs PR here: https://github.com/hashicorp/nomad/pull/24275

**Comment by neuroserve** [comment]
I'm pretty sure, that there was/is a reason to set token locality to local. 

What would be the "right" way in such an environment? I suspect "workload identities" - but there I'm missing more docs on how to set this up in a production grade manner (nginx proxy for JWT access, one Nomad service in consul - not three, etc.).

The docs for consul service mesh and mesh-gateways from Hashicorp are pretty good (and there's plenty more from other sources on that topic). Docs on workload identities are a little bit scarce in that respect.

Kind regards,
Töns

**Comment by tgross** [comment]
@neuroserve did you read @rboyer's comment here: https://github.com/hashicorp/consul/issues/21863#issuecomment-2430263578 ?

**Comment by neuroserve** [comment]
OK. I had

```
service_prefix "" {
    policy = "read"
}

node_prefix "" {
    policy = "read"
}

query_prefix "" {
    policy = "read"
}

```

in my anonymous policy. I have changed to
```

service_prefix "*" {
    policy = "read"
}

node_prefix "*" {
    policy = "read"
}

query_prefix "" {
    policy = "read"
}

```

now and will have to check, whether something has changed.

Thank you for the hint.

**Comment by neuroserve** [comment]
Just to add the result here: Even changing to a anonymous policy


```
service_prefix "*" {
    policy = "read"
}

node_prefix "*" {
    policy = "read"
}

query_prefix "" {
    policy = "read"
}

```
for the anonymous token in Consul did change the error. Still reporting 
"failed to derive SI token: Unexpected response code: 500 (rpc error making call: Local tokens are disabled)"

Still no idea how cross-datacenter service mesh with nomad + consul should work.
